### PR TITLE
access_log: metadata formatter extension

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
  Common metadata formatter extension to handle all types of metadata.

Additional Description:
This PR addresses several issues:
1. MetadataFormatter base class takes a pointer to a function to access specific type of metadata. All other methods operating on internals of metadata structure are moved to base class. This simplified adding new types of metadata.
2. new type of  access_log METADATA tag is introduced to handle all types of metadata. It is implemented as formatter extension. It handles existing tags DYNAMIC_METADATA and CLUSTER_METADATA, so two ways of accessing dynamic and cluster metadata are allowed now. Maybe we should obsolete DYNAMIC_METADATA and CLUSTER_METADATA in the future?
3. Adds handler for ROUTE metadata. See issue #3006.

Follow up to #16121

Risk Level: Low
Testing: Unit tests.
Docs Changes: Yes.
Release Notes: Yes.
Platform Specific Features: No

### CODE REVIEW

File: source/extensions/formatter/metadata/metadata.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+private:
+  static const size_t MetadataFormatterParamStart{sizeof("METADATA(") - 1};
+
+  // Map used to dispatch types of metadata to individual handlers which will
+  // access required metadata object.
+  std::map<std::string,
+           std::function<::Envoy::Formatter::FormatterProviderPtr(
+               const std::string& filter_namespace, const std::vector<std::string>& path,
+               absl::optional<size_t> max_length)>>
```
Comment (rgs1) [inline source/extensions/formatter/metadata/metadata.h:29]: nit: 

```
using FormatterProvider = std::function<::Envoy::Formatter::FormatterProviderPtr(const std::string& \
filter_namespace, const std::vector<std::string>& path, absl::optional<size_t> max_length)>;
std::map<std::string, FormatterProvider> metadata_formatter_providers_;
```

File: source/extensions/formatter/metadata/metadata.cc [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      : ::Envoy::Formatter::MetadataFormatter(filter_namespace, path, max_length,
+                                              [](const StreamInfo::StreamInfo& stream_info)
+                                                  -> const envoy::config::core::v3::Metadata* {
+                                                auto route = stream_info.routeEntry();
+
+                                                if (route == nullptr) {
+                                                  return nullptr;
+                                                }
+                                                return &route->metadata();
+                                              }) {}
```
Comment (rgs1) [inline source/extensions/formatter/metadata/metadata.cc:29]: somewhat weird formatting, but it's probably fine and it's not worth creating the lambda from a private method.

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+//    where TYPE is type of metadata (see above for supported types),
+//    NAMESPACE is the filter namespace used when setting the metadata, KEY is an optional
+//    lookup up key in the namespace with the option of specifying nested keys separated by ':',
+//    and Z is an optional parameter denoting string truncation up to Z characters long.
+//    The data will be logged as a JSON string. For example, for the following ROUTE metadata:
+//
+//    ``com.test.my_filter: {"test_key": "foo", "test_object": {"inner_key": "bar"}}``
+//
+//    * %METADATA(ROUTE:com.test.my_filter)% will log: ``{"test_key": "foo", "test_object": {"inner_key": "bar"}}``
+//    * %METADATA(ROUTE:com.test.my_filter:test_key)% will log: ``"foo"``
```
Comment (rgs1) [inline api/envoy/extensions/formatter/metadata/v3/metadata.proto:36]: hmm it should be `foo` without the quotes since #16736?

**Review by rgs1 — COMMENTED** [review]
LGTM, thanks. Just a few nits and one question.

File: source/extensions/formatter/metadata/metadata.h (hunk 1 above)
Comment (cpakulski) [reply source/extensions/formatter/metadata/metadata.h:29]: Thanks. It is cleaner now.

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto (hunk 3 above)
Comment (cpakulski) [reply api/envoy/extensions/formatter/metadata/v3/metadata.proto:36]: I blindly copied it from https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage.html. Corrected docs everywhere. Thanks for catching it.

**Review by rgs1 — APPROVED** [review]
LGTM, thanks.

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+//    * %METADATA(ROUTE:com.unknown_filter)% will log: ``-``
+//    * %METADATA(ROUTE:com.test.my_filter:unknown_key)% will log: ``-``
+//    * %METADATA(ROUTE:com.test.my_filter):25% will log (truncation at 25 characters): ``{"test_key": "foo", "test``
+//
+//  .. note::
+//
+//    For typed JSON logs, this operator renders a single value with string, numeric, or boolean type
+//    when the referenced key is a simple value. If the referenced key is a struct or list value, a
+//    JSON struct or list is rendered. Structs and lists may be nested. In any event, the maximum
+//    length is ignored
```
Comment (adisuissa) [inline api/envoy/extensions/formatter/metadata/v3/metadata.proto:48]: nit: Line should end with a period (.)

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+//    * %METADATA(ROUTE:com.test.my_filter:test_object:inner_key)% will log: ``bar``
+//    * %METADATA(ROUTE:com.unknown_filter)% will log: ``-``
+//    * %METADATA(ROUTE:com.test.my_filter:unknown_key)% will log: ``-``
+//    * %METADATA(ROUTE:com.test.my_filter):25% will log (truncation at 25 characters): ``{"test_key": "foo", "test``
+//
+//  .. note::
+//
+//    For typed JSON logs, this operator renders a single value with string, numeric, or boolean type
+//    when the referenced key is a simple value. If the referenced key is a struct or list value, a
+//    JSON struct or list is rendered. Structs and lists may be nested. In any event, the maximum
```
Comment (adisuissa) [inline api/envoy/extensions/formatter/metadata/v3/metadata.proto:47]: Question about not adhering to the max length: can someone use this to create a buffer overflow?

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+//    length is ignored
+//
+//  .. note::
+//
+//    METADATA(DYNAMIC:NAMESPACE:KEY):Z is equivalent to :ref:`DYNAMIC_METADATA(NAMESPACE:KEY):Z<config_access_log_format_dynamic_metadata>`
+//    METADATA(CLUSTER:NAMESPACE:KEY):Z is equivalent to :ref:`CLUSTER_METADATA(NAMASPACE:KEY):Z<config_access_log_format_cluster_metadata>`
```
Comment (adisuissa) [inline api/envoy/extensions/formatter/metadata/v3/metadata.proto:53]: Re your question, I think it makes sense that the previous `DYNAMIC_METADATA` and `CLUSTER_METADATA` should be deprecated, and the new approach should be encouraged, as it is more general.
@htuch WDYT?

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
Overall API looks good, left a question.

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto (hunk 5 above)
Comment (cpakulski) [reply api/envoy/extensions/formatter/metadata/v3/metadata.proto:47]: I checked the code how _max_length_ is used. It does not allocate any buffer but truncates a string to be displayed. So, I think there is no risk for buffer overflow.

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto (hunk 4 above)
Comment (cpakulski) [reply api/envoy/extensions/formatter/metadata/v3/metadata.proto]: Fixed. Thanks for catching it!

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto (hunk 6 above)
Comment (htuch) [reply api/envoy/extensions/formatter/metadata/v3/metadata.proto:53]: I think so. We have more metadata than even those, so having a more general purpose mechanism makes sense to me.

**Comment by cpakulski** [comment]
@adisuissa Would you mind doing LGTM for API change? Thanks!

**Comment by adisuissa** [comment]
> @adisuissa Would you mind doing LGTM for API change? Thanks!

Is there a path to deprecating `DYNAMIC_METADATA` and `CLUSTER_METADATA`?

**Comment by cpakulski** [comment]
I suggest the following deprecation path:
1. I will update `DYNAMIC_METADATA` and `CLUSTER_METADATA` documentation indicating that `METADATA` should be used and that both of those tags will be removed in the future. This will be done in this PR.
2. In the next release I will add runtime overwrite to use `DYNAMIC_METADATA` and `CLUSTER_METADATA` tags. They will be disabled by default.
3. After another 1-2 releases  from adding runtime overwrite `DYNAMIC_METADATA` and `CLUSTER_METADATA` will be removed permanently.

File: api/envoy/extensions/formatter/metadata/v3/metadata.proto (hunk 6 above)
Comment (lizan) [reply api/envoy/extensions/formatter/metadata/v3/metadata.proto:53]: +1

File: source/extensions/formatter/metadata/metadata.cc [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  metadata_formatter_providers_["ROUTE"] = [](const std::string& filter_namespace,
+                                              const std::vector<std::string>& path,
+                                              absl::optional<size_t> max_length) {
+    return std::make_unique<RouteMetadataFormatter>(filter_namespace, path, max_length);
+  };
+}
+
+::Envoy::Formatter::FormatterProviderPtr
+MetadataFormatterCommandParser::parse(const std::string& token, size_t, size_t) const {
+  const std::string METADATA_TOKEN = "METADATA(";
```
Comment (lizan) [inline source/extensions/formatter/metadata/metadata.cc:56]: nit: constexpr absl::string_view

File: source/extensions/formatter/metadata/metadata.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+// Access log handler for METADATA( command.
+class MetadataFormatterCommandParser : public ::Envoy::Formatter::CommandParser {
+public:
+  MetadataFormatterCommandParser();
+  ::Envoy::Formatter::FormatterProviderPtr parse(const std::string& token, size_t,
+                                                 size_t) const override;
+
+private:
+  static const size_t MetadataFormatterParamStart{sizeof("METADATA(") - 1};
```
Comment (lizan) [inline source/extensions/formatter/metadata/metadata.h:22]: not used

**Comment by adisuissa** [comment]
> I suggest the following deprecation path:
> 
> 1. I will update `DYNAMIC_METADATA` and `CLUSTER_METADATA` documentation indicating that `METADATA` should be used and that both of those tags will be removed in the future. This will be done in this PR.
> 2. In the next release I will add runtime overwrite to use `DYNAMIC_METADATA` and `CLUSTER_METADATA` tags. They will be disabled by default.
> 3. After another 1-2 releases  from adding runtime overwrite `DYNAMIC_METADATA` and `CLUSTER_METADATA` will be removed permanently.

SGTM, and thanks for doing this!
Let me know once (1) is done, so I can take another look and approve the API.

File: source/extensions/formatter/metadata/metadata.h (hunk 8 above)
Comment (cpakulski) [reply source/extensions/formatter/metadata/metadata.h]: Fixed. Thanks for catching it.

File: source/extensions/formatter/metadata/metadata.cc (hunk 7 above)
Comment (cpakulski) [reply source/extensions/formatter/metadata/metadata.cc]: Corrected. Thanks.

**Comment by cpakulski** [comment]
@adisuissa Added notes about deprecation in the future releases.

I will be away for few weeks. If there are any additional comments on this PR, I will address them when I get back.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by htuch** [comment]
@cpakulski we'll need 12 months of support from when it's deprecated, but plan sounds fine (assuming appropriate signaling via the version history).

**Comment by htuch** [comment]
See https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md#breaking-change-policy

**Comment by cpakulski** [comment]
@htuch Thanks for clarifying. I think the code and docs (there is a note stating that METADATA tag should be used instead if DYNAMIC and CLUSTER_METADATA) are good right now. After 1.20 is released I will add a log warning and increment _runtime.deprecated_feature_use_.

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
It looks like all comments were addressed. LGTM. Waiting for last pass from @lizan

**Review by lizan — APPROVED** [review]

**Pull request merged**